### PR TITLE
Create API

### PR DESCRIPTION
**Pull request by anderCM**
## Requirements

- Set up the repository/repositories on GitHub and use Gitflow.
- You should use Postgres as your database
- Use Rails to create backend API.
- The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.
-  Add full documentation for your API.
- Add validations.
- Add unit tests for models and controllers.
- Validate parameters in controllers, models and migrations.

#### Front-end application PR
- CLICK [HERE](https://github.com/limavfabio/VenomPrecision-front-end/pull/31)

### CODE REVIEW

File: app/controllers/api/v1/reservations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    if @reservation.update(reservation_params)
+      render json: @reservation
+    else
+      render json: @reservation.errors, status: :unprocessable_entity
+    end
+  end
+
+  # DELETE /reservations/1
+  def destroy
+    @reservation.destroy
```
Comment (mcrd25) [inline app/controllers/api/v1/reservations_controller.rb:45]: What happens if it is not destroyed successfully?

File: app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    if @user.update(user_params)
+      render json: @user
+    else
+      render json: @user.errors, status: :unprocessable_entity
+    end
+  end
+
+  # DELETE /users/1
+  def destroy
+    @user.destroy
```
Comment (mcrd25) [inline app/controllers/api/v1/users_controller.rb:38]: What happens if it is not destroyed successfully?

**Review by mcrd25 — CHANGES_REQUESTED** [review]
Hi Team,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- You seem to have met all functional requirements 👍 
- Have API documentation 👍 
- Implemented routes well 👍 
- Your code is clean enough 👍 

## Required Changes ♻️
_Check the comments under the review._
- Make sure to ensure all tests are passing before code reviews, if this is something you need extra support with, please do let me know

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [OPTIONAL] removing the `test` directory since you are using `rspec` for testing and not native testing.
- [OPTIONAL] Consider using the `timecop` gem to freeze time when running tests. It can help avoid errors: https://github.com/travisjeffery/timecop
- [OPTIONAL] Should always have positive and negative tests to ensure it works the way you expect it. Eg. What do you expect the output to be when it's not valid or when valid etc.
Consider making all your tests more extensive, not just this one. For this simple application, it may not catch a possible crippling bug, but in bigger applications, it definitely can. Consider this for all projects moving forward.


**Hope all is well, and that my comments were helpful. 😸**

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/reservation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  it 'should be stored in DB' do
+    user = User.create(username: 'andercm')
+    product = Product.create(name: 'Gun 1', image: 'https://urlpicture.com', price: 27.20, owner: user)
+
+    reservation = Reservation.create(date: '2023-06-16', user:, product:)
+
+    expect(reservation.persisted?).to eq(true)
+  end
```
[the remark this answers is not in the document]
Comment (chelobotix) [reply spec/models/reservation_spec.rb:30]: Hello @mcrd25, the problem has been fixed.
Thank you

File: spec/integration/api/v1/reservations_spec.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          example.metadata[:response][:content] = {
+            'application/json' => {
+              example: JSON.parse(response.body, symbolize_names: true)
+            }
+          }
+        end
+        run_test!
+      end
+    end
+  end
```
[the remark this answers is not in the document]
Comment (chelobotix) [reply spec/integration/api/v1/reservations_spec.rb:107]: @mcrd25, just wanted to give you a heads-up that the integration folder tests were automatically created by Rswag to help with API-Documentation. For the correct type of request tests, you can find them in the Request folder. Thanks!

File: spec/integration/api/v1/products_spec.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            'application/json' => {
+              example: JSON.parse(response.body, symbolize_names: true)
+            }
+          }
+        end
+        run_test!
+      end
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (chelobotix) [reply spec/integration/api/v1/products_spec.rb:94]: @mcrd25, just wanted to give you a heads-up that the integration folder tests were automatically created by Rswag to help with API-Documentation. For the correct type of request tests, you can find them in the Request folder. Thanks!

File: spec/integration/api/v1/users_spec.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+          example.metadata[:response][:content] = {
+            'application/json' => {
+              example: JSON.parse(response.body, symbolize_names: true)
+            }
+          }
+        end
+        run_test!
+      end
+    end
+  end
```
[the remark this answers is not in the document]
Comment (chelobotix) [reply spec/integration/api/v1/users_spec.rb:95]: @mcrd25, just wanted to give you a heads-up that the integration folder tests were automatically created by Rswag to help with API-Documentation. For the correct type of request tests, you can find them in the Request folder. Thanks!

File: spec/requests/api/v1/product_spec.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      json_response = JSON.parse(response.body)
+      expect(response).to have_http_status(:success)
+      expect(json_response['name']).to eq('Tippmann Cronus Paintball Gun')
+    end
+  end
+
+  describe 'DESTROY api/v1/products' do
+    it 'return http success' do
+      delete api_v1_product_path(@product1.id)
+      expect(response).to have_http_status(:success)
```
[the remark this answers is not in the document]
Comment (chelobotix) [reply spec/requests/api/v1/product_spec.rb]: Hello @mcrd25, the problem has been fixed.
Thank you

File: spec/requests/api/v1/reservation_spec.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      expect(json_response['date']).to eq('2023-05-04')
+    end
+  end
+
+  describe 'DESTROY api/v1/users/reservation' do
+    it 'return http success' do
+      delete api_v1_user_reservation_path(user_id: @user1.id, id: @reservation1.id)
+      expect(response).to have_http_status(:success)
+    end
+  end
```
[the remark this answers is not in the document]
Comment (chelobotix) [reply spec/requests/api/v1/reservation_spec.rb]: Hello @mcrd25, the problem has been fixed.
Thank you

File: app/controllers/api/v1/users_controller.rb (hunk 2 above)
Comment (anderCM) [reply app/controllers/api/v1/users_controller.rb]: @mcrd25  We have added a validation and returning a custom message if the user is destroyed
```
  def destroy
    if @user.destroy
      render json: { status: 'success', message: 'The user has been deleted' }
    else
      render json: @user.errors, status: :unprocessable_entity
    end
  end
  ```

File: app/controllers/api/v1/reservations_controller.rb (hunk 1 above)
Comment (anderCM) [reply app/controllers/api/v1/reservations_controller.rb]: @mcrd25  We have added a validation and returning a custom message if the reservation is destroyed
```
  def destroy
    if @reservation.destroy
      render json: { status: 'success', message: 'The reservation has been deleted' }
    else
      render json: @reservation.errors, status: :unprocessable_entity
    end
  end
  ```

File: spec/requests/api/v1/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    end
+  end
+
+  describe 'DESTROY api/v1/products' do
+    it 'return http success' do
+      delete api_v1_user_path(@user1.id)
+      expect(response).to have_http_status(:success)
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (anderCM) [reply spec/requests/api/v1/user_spec.rb]: @mcrd25  We have fixed these issues, now all tests are passing

File: README.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+rails server
+```
+
+### Run tests
+
+To run tests, run the following command:
+
+```sh
+rspec spec
+```
```
Comment (Meltrust) [inline README.md:124]: - [ ] OPTIONAL: You found a way to exclude the swagger files from the rspec command 👍 .  

You can use/add this other command to your instructions If you want your users to get a green pass in the terminal for rswag tests:

```
rake rswag
```

Result:

image

**Review by Meltrust — APPROVED** [review]
#### Hi team,


# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- Failure/error handling for `destroy` has been implemented ✔️ 
- Tests are now passing✔️ 
- rswag generated files have been excluded from the rspec command avoiding questionable test runs✔️ 
- Great job✔️ 



Cheers and Happy coding!👏👏👏


------

**Pull request merged**